### PR TITLE
add support for TinyDB feature provider

### LINKED ISSUE #1723: add TinyDB feature provider
**Is your feature request related to a problem? Please describe.**
Enhance the TinyDBCatalogue Provider to additionally support features.

**Describe the solution you'd like**
Provision of a TinyDB provider to publish feature data.

### PR DESCRIPTION
**Pull request by tomkralidis**
# Overview
Add support for TinyDB as a feature provider.

Additionally updates CITE test setup and docs.

# Related Issue / discussion
Fixes #1723 

# Additional information

# Dependency policy (RFC2)

- [x] I have ensured that this PR meets [RFC2](https://pygeoapi.io/development/rfc/2) requirements

# Updates to public demo

- [x] I have ensured that breaking changes to the [pygeoapi master demo server](https://github.com/geopython/demo.pygeoapi.io) have been addressed
  - [ ] https://github.com/geopython/demo.pygeoapi.io/blob/master/services/pygeoapi_master/local.config.yml

# Contributions and licensing

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [x] I'd like to contribute [feature X|bugfix Y|docs|something else] to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution
- [x] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines

### CODE REVIEW

File: tests/test_tinydb_provider.py
Comment (webb-ben) [inline tests/test_tinydb_provider.py:5]: 2024?

**Review by webb-ben — APPROVED** [review]

**Comment by tomkralidis** [comment]
FYI the associated CITE instance [PR](https://github.com/geopython/demo.pygeoapi.io/pull/58) can be merged if/once this is approved/merge.

File: tests/test_tinydb_provider.py
Comment (tomkralidis) [reply tests/test_tinydb_provider.py]: Good catch!  Updated here and in `pygeoapi/plugin.py`.

File: docs/source/data-publishing/ogcapi-features.rst
Comment (justb4) [inline docs/source/data-publishing/ogcapi-features.rst:34]: align comma in table for TinyDB

File: pygeoapi/plugin.py
Comment (justb4) [inline pygeoapi/plugin.py:63]: Looks like `TinyDBCatalogueProvider` has been renamed to `TinyDBProvider` ? So can be removed here or for forward compat map to `TinyDBProvider`...

Comment (justb4) [reply pygeoapi/plugin.py:63]: (ok I see that both are defined).

File: pygeoapi/provider/tinydb_.py
Comment (justb4) [inline pygeoapi/provider/tinydb_.py:1]: There are quite some `if self._catalogue` statements in the base class `TinyDBProvider`, while at the same time there is a subclass `TinyDBCatalogueProvider` . My suggestion is to refactor those conditionals into the subclass.

**Review by justb4 — CHANGES_REQUESTED** [review]
Two main comments:
- refactor Catalogue-specific statements into subclass `TinyDBCatalogueProvider`
- add also unit tests for `TinyDBCatalogueProvider`

File: docs/source/data-publishing/ogcapi-features.rst
Comment (tomkralidis) [reply docs/source/data-publishing/ogcapi-features.rst]: Fixed.

File: pygeoapi/plugin.py
Comment (tomkralidis) [reply pygeoapi/plugin.py:63]: Yes, there are nuances between a feature and record provider here.

File: pygeoapi/provider/tinydb_.py
Comment (tomkralidis) [reply pygeoapi/provider/tinydb_.py:1]: That was my original implementation, but I decided to add this conditional (3x) in the code so as not to have to loop after search results which would affect performance.

Comment (justb4) [reply pygeoapi/provider/tinydb_.py:1]: (maybe not use return values, but like: 
```
    def _add_text_search_field(self, fields):
        fields['q'] = {'type': 'string'}
        // return fields
```

**Comment by tomkralidis** [comment]
> Two main comments:
> 
>     * refactor Catalogue-specific statements into subclass `TinyDBCatalogueProvider`
> 
>     * add also unit tests for `TinyDBCatalogueProvider`

Note that these already exist in `tests/test_tinydb_catalogue_provider.py`.

**Comment by tomkralidis** [comment]
@justb4 PR updated, for review.

File: pygeoapi/provider/tinydb_.py
Comment (justb4) [inline pygeoapi/provider/tinydb_.py:350]: Major: IMO will not have desired effect: `self.fields = self.get_fields()` is already called in the base class constructor. Or `self.fields = self.get_fields()` should be called again, but see my suggestion to make `_excludes` a class var as it is in effect a property of the classes.

Comment (justb4) [reply pygeoapi/provider/tinydb_.py:350]: To consider: `self.fields` is never referenced after assignment so maybe `self.fields = self.get_fields()` can be removed and the above `self._excludes =` -statement will be ok. `get_fields()` is a Provider base.py method and the way to access fields I guess.

**Review by justb4 — CHANGES_REQUESTED** [review]
See inline comments: think all will be resolved by simply removing `self.fields = self.get_fields()` from constructor.

File: pygeoapi/provider/tinydb_.py
Comment (tomkralidis) [reply pygeoapi/provider/tinydb_.py:350]: If we remove `self.fields = self.get_fields()` from the constructor then this will affect other parts of `pygeoapi.api.itemtypes`.  Suggest we leave this as is in this PR and address in a subsequent PR later on.

Comment (justb4) [reply pygeoapi/provider/tinydb_.py:350]: Ok, then the only remaining is IMO to resolve the proper `_excludes`.

Comment (justb4) [reply pygeoapi/provider/tinydb_.py:350]: Do you agree that the current order will not work? Hence my suggestion for a class var (see my example) or another solution?

**Review by justb4 — CHANGES_REQUESTED** [review]

File: pygeoapi/provider/tinydb_.py
Comment (tomkralidis) [reply pygeoapi/provider/tinydb_.py:350]: The default TinyDB provider as a result of this PR has an empty `_excludes` member.  The TinyDBCatalogue provider sets `_excludes` to remove a given property that is a pygeoapi "extra".

**Comment by tomkralidis** [comment]
@justb4 this order works and accomplishes the need to remove the specified fields from catalogue responses.

File: pygeoapi/provider/tinydb_.py
Comment (webb-ben) [reply pygeoapi/provider/tinydb_.py:350]: fwiw other providers make the call to `self.get_fields()` at the end of the init method which will either fetch self.fields if set, or set self.fields.

[esri.py](https://github.com/geopython/pygeoapi/blob/e8ce02167fd670bb515f8216d699a3e9abca21d4/pygeoapi/provider/esri.py#L70)

[sqlite.py](https://github.com/geopython/pygeoapi/blob/e8ce02167fd670bb515f8216d699a3e9abca21d4/pygeoapi/provider/sqlite.py#L82)

Happy to establish a normative behavior between the two peoples `self.fields = self.get_fields()` is present in

[postgresql.py](https://github.com/geopython/pygeoapi/blob/e8ce02167fd670bb515f8216d699a3e9abca21d4/pygeoapi/provider/postgresql.py#L127)

[elasticsearch_.py](https://github.com/geopython/pygeoapi/blob/e8ce02167fd670bb515f8216d699a3e9abca21d4/pygeoapi/provider/elasticsearch_.py#L90)

**Comment by justb4** [comment]
> @justb4 this order works and accomplishes the need to remove the specified fields from catalogue responses.

Hmm, I still don't think so. Here is a similar example I used to investigate, showing both the instance and class var:

```
class Base:
    # class variable
    class_var = ['Base']
    def __init__(self):
        self.exclude = 'Exclude by Base'
        print('Parent init')
        self.show_info()
    def show_info(self):
        # Accessing class variable
        print("Base: self.exclude=", self.exclude)
        print("Base: Base.class_var=", Base.class_var)

class Derived(Base):

    def __init__(self):
        print('Derived init')
        Base.class_var.append('Derived')
        super().__init__()
        self.exclude = 'Exclude by Derived'

    def show_info(self):
        # Accessing class variable
        super().show_info()
        print("Derived: self.exclude=", self.exclude)
        print("Derived: Base.class_var=", Base.class_var)

derived = Derived()
```
`self.exclude = 'Exclude by Derived'` is called after the Base constructor. Guess the outcome...

File: pygeoapi/provider/tinydb_.py
Comment (tomkralidis) [reply pygeoapi/provider/tinydb_.py:350]: Good points.  Suggest we address this as part of another PR

**Comment by tomkralidis** [comment]
@justb4 feel free to test the branch with/without the exclude in the TinyDBCatalogue class?

**Comment by justb4** [comment]
Ok, all tests pass. But makes sense as `p.get_fields()` is called. If I make
```
def test_query(config):
    p = TinyDBCatalogueProvider(config)

    fields = p.fields
```
as it results from constructor, the field is not excluded and test fails:

```
(pygeoapi-3.10.12) just@savu:~/project/pygeoapi/pygeoapi.git$ pytest tests/test_tinydb_catalogue_provider.py 
========================================================= test session starts =========================================================
platform darwin -- Python 3.10.12, pytest-8.2.2, pluggy-1.5.0
rootdir: /Users/just/project/pygeoapi/pygeoapi.git
configfile: pytest.ini
plugins: Faker-26.0.0, cov-5.0.0, env-1.1.3, anyio-4.4.0
collected 5 items                                                                                                                     

tests/test_tinydb_catalogue_provider.py F....                                                                                   [100%]

============================================================== FAILURES ===============================================================
_____________________________________________________________ test_query ______________________________________________________________

config = {'data': PosixPath('/private/var/folders/89/ggkklkwd5d7g9sw3pfbqwcvm0000gn/T/pytest-of-just/pytest-1/test_query0/sample-records.tinydb'), 'id_field': 'externalId', 'name': 'TinyDBCatalogue', 'time_field': 'created', ...}

    def test_query(config):
        p = TinyDBCatalogueProvider(config)
    
        fields = p.fields
>       assert len(fields) == 9
E       AssertionError: assert 10 == 9
E        +  where 10 = len({'_metadata-anytext': {'type': 'string'}, 'created': {'format': 'date-time', 'type': 'string'}, 'description': {'type': 'string'}, 'externalIds': {'type': 'string'}, ...})

tests/test_tinydb_catalogue_provider.py:97: AssertionError
======================================================= short test summary info =======================================================
FAILED tests/test_tinydb_catalogue_provider.py::test_query - AssertionError: assert 10 == 9
===================================================== 1 failed, 4 passed in 0.10s =====================================================
(py`
```
'_metadata-anytext' is _not_ excluded... as can be expected.

**Comment by tomkralidis** [comment]
What if we normalized all access via `get_fields()`, and remove the `.fields` property?  Again, this would be part of another PR.

**Comment by justb4** [comment]
Function access is always to be preferred over exposing internal instance vars.

But harder to assess (performance) impact. As for some Providers it may mean accessing a remote source like a DB or even WFS. But then again `get_fields()` may smart-cache. Could even be implemented once in Base class e.g. with `@property` ...

**Comment by tomkralidis** [comment]
> \\@property
>     def fields(self):
>         return self.get_fields()

I guess you mean:

```python
@property
def fields(self):
    return self.get_fields()
```

Suggest we address as part of another PR (given the scope of this PR is to expand TinyDB support).  While we've identified some inconsistency/issue with `.fields`/`get_fields()`, this is better address in a separate PR given it will be a substantial change across numerous feature/record providers and their calling code.

@justb4 works for you?

**Comment by justb4** [comment]
Ok, yes, but in my code I meant to cache the fields on first access (or leave that to each provider). Still in the current state `TinyDBCatalogueProvider` self.fields is incorrect. But don't know how much impact that has.

**Comment by tomkralidis** [comment]
> ll in the current state `TinyDBCatalogueProvider` self.fields is incorrect. But don't know how much impact that has.

I think there is no impact in the context of the codepath and how the plugin is implemented.

**Review by justb4 — APPROVED** [review]
Ok, we'll fix 'excluds' issue later. For sake of progress.

**Review by doublebyte1 — COMMENTED** [review]
Nice job!

**Pull request merged**
